type-inference-in-readr-and-arrow.rmd

```diff
@@ -0,0 +1,271 @@
+
+```{r}
+int_or_dbl <- data.frame(
+  x = c(1L, 2L, 3L)
+)
+
+write.csv(int_or_dbl, "int_or_dbl.csv", row.names = FALSE)
+
+readLines("int_or_dbl.csv")
+
+# double
+readr::read_csv("int_or_dbl.csv")
+```
+
+```{r}
+
+# integer
+readr::read_csv("int_or_dbl.csv", col_types = list(.default = col_character())) %>%
+  type_convert(guess_integer = TRUE)
+
+```
+
+```{r}
+# integer
+arrow::read_csv_arrow("int_or_dbl.csv")
+```
+
+### 32-bit integers
+
+Another difference between readr and arrow is the difference between how integers larger than 32 bits are read in.  Natively, R can only support 32-bit integers, though it can support 64-bit integers via the bit64 package.  If we create a CSVe with one column containing the largest integer that R can natively support, and then another column containing that value plus 1, we get different behaviour when we import this data with readr and arrow.  In readr, when we enable integer guessing, the smaller value is read in as an integer, and the larger value is read in as a double.  In arrow, the use of the bit64 package means both values can be represented as integers.
+
+```{r}
+sixty_four <- data.frame(x = 2^31 - 1, y = 2^31)
+
+write.csv(sixty_four, "sixty_four.csv", row.names = FALSE)
+
+# integer by default
+readr::read_csv("sixty_four.csv")
+```
+
+```{r}
+# 32 or 64 bit integer depending on value size
+arrow::read_csv_arrow("sixty_four.csv")
+```
+
+```{r}
+# 32 bit integer or double depending on value size
+readr::read_csv("sixty_four.csv", col_types = list(.default = col_character())) %>%
+  type_convert(guess_integer = TRUE)
+```
+
+### The "number" type
+
+One really cool feature in readr is the "number" type.  This allows reading in of values which have been stored as character data with commas to separate the thousands as doubles.  This is not supported in arrow.
+
+```{r}
+number_type <- data.frame(
+  x = c("1,000", "1,250")
+)
+
+write.csv(number_type, "number_type.csv", row.names = FALSE)
+
+# double type, but read in as number initially
+readr::read_csv("number_type.csv")
+```
+
+```{r}
+# read in as character data in Arrow
+arrow::read_csv_arrow("number_type.csv")
+```
+
+### Dictionaries/Factors
+
+Anyone who’s been around long enough might remember that R’s native CSV reading function `read.csv()` had a default setting of importing character columns as factors (I definitely have `read.csv(..., stringAsFactors=FALSE)` carved into a groove in some dark corner of my memory).  This default was changed in version 4.0.0, released in April 2020, reflecting the fact that in most cases users want their string data to be imported as characters unless otherwise specified.  Still, some datasets contain character data which users do want to import as factors.  In reader, this can be controlled by manually specifying the column as a factor
```

Comment:
```suggestion
Anyone who’s been around long enough might remember that R’s native CSV reading function `read.csv()` had a default setting of importing character columns as factors (I definitely have `read.csv(..., stringAsFactors=FALSE)` carved into a groove in some dark corner of my memory).  This default was changed in version 4.0.0, released in April 2020, reflecting the fact that in most cases users want their string data to be imported as characters unless otherwise specified.  Still, some datasets contain character data which users do want to import as factors.  In readr, this can be controlled by manually specifying the column as a factor
```